src/vcpkg/base/files.cpp

```diff
@@ -1682,6 +1824,206 @@ namespace vcpkg
         {
             return get_regular_files_impl<stdfs::directory_iterator>(dir, ec);
         }
+#else // ^^^ _WIN32 // !_WIN32 vvv
+#if defined(_DIRENT_HAVE_D_TYPE)
+#define VCPKG_GET_D_TYPE(dirent) ((dirent)->d_type)
+#define VCPKG_DT_UNKNOWN DT_UNKNOWN
+#define VCPKG_DT_REG DT_REG
+#define VCPKG_DT_DIR DT_DIR
+#else
+#define VCPKG_GET_D_TYPE(dirent) 0
+#define VCPKG_DT_UNKNOWN 0
+#define VCPKG_DT_REG 1
+#define VCPKG_DT_DIR 2
+#endif
+
+        // Selector is a function taking (unsigned char dtype, mode_t mode) and returning bool
+        // dtype is the value from the d_type member of struct dirent. If the system does not support the d_type
+        // member, this will always be set to VCPKG_DT_UNKNOWN.
+        // mode is the st_mode member of struct stat. It is populated if and only if dtype is VCPKG_DT_UNKNOWN.
+        // Note that many file systems always set dtype to DT_UNKNOWN, so the lstat fallback must exist even on systems
+        // which have the d_type member
+        template<class Selector>
+        static void get_files_recursive_impl(std::vector<Path>& result,
+                                             const Path& base,
+                                             std::error_code& ec,
```

Comment:
I think the error_code in the recursive case needs to be paired with an out-var for the failing path; otherwise the error will be meaninglessly out of context.

Reply:
That is already the status quo for this interface. We can talk about potentially changing that interface but I don't think that's this PR.